lib/propolis/src/hw/virtio/p9fs.rs

```diff
@@ -974,3 +950,54 @@ pub(crate) fn write_error(ecode: u32, chain: &mut Chain, mem: &MemCtx) {
     let buf = out.as_mut_slice();
     write_buf(buf, chain, mem);
 }
+
+pub(crate) fn p9_write_file(
+    file: &impl std::os::fd::AsRawFd,
+    chain: &mut Chain,
+    mem: &MemCtx,
+    count: usize,
+    offset: i64,
+) {
+    // Form the rread header. Unfortunately we can't do this with the Rread
+    // structure because the count is baked into the data field which is tied
+    // to the length of the vector and filling that vector is what we're
+    // explicitly trying to avoid here.
+    let sz = size_of::<u32>() + // size
+            size_of::<u8>()  + // typ
+            size_of::<u16>() + // tag
+            size_of::<u32>() +  // data.count
+            count; // data
+    let mut header = Vec::with_capacity(11);
+    header.extend_from_slice(&(sz as u32).to_le_bytes());
+    header.push(p9ds::proto::MessageType::Rread as u8);
+    header.extend_from_slice(&0u16.to_le_bytes());
+    header.extend_from_slice(&(count as u32).to_le_bytes());
+
+    // Send the header to the guest from the buffer constructed above. Then
+    // send the actual file data
+    let mut header_done = false;
+    let mut done = 0;
+    let _total = chain.for_remaining_type(false, |addr, len| {
+        let mut remain = len;
+        let mut copied = 0;
+        if !header_done {
+            mem.write_from(addr, &header, header.len()).unwrap();
+            header_done = true;
+            remain -= header.len();
+            copied += header.len();
+        }
+        let addr = GuestAddr(addr.0 + copied as u64);
```

Comment:
Could use `GuestAddr::offset()` for this

Reply:
Done